hbase-client/src/main/java/org/apache/hadoop/hbase/client/ConnectionImplementation.java

```diff
@@ -382,6 +390,24 @@ public void newDead(ServerName sn) {
       default:
         // Doing nothing
     }
+
+    this.cachedMasterStateSupplier = Suppliers.memoizeWithExpiration(() -> {
+      if (this.masterServiceState.getStub() == null) {
+        return false;
+      }
+      try {
+        LOG.info("Getting master state using rpc call");
+        return this.masterServiceState.isMasterRunning();
+      } catch (UndeclaredThrowableException e) {
+        // It's somehow messy, but we can receive exceptions such as
+        // java.net.ConnectException but they're not declared. So we catch it...
+        LOG.info("Master connection is not running anymore", e.getUndeclaredThrowable());
+        return false;
+      } catch (IOException se) {
+        LOG.warn("Checking master connection", se);
+        return false;
+      }
+    }, conf.getLong(MASTER_STATE_CACHE_TIMEOUT_SEC, 30), TimeUnit.SECONDS);
```

Comment:
IMO we should default opt-out this feature without default to 0, then also add a unit test to a connection related test to have the the timeout set to 30 and see if anything breaks.

Reply:
Hi @taklwu I have implemented the review comments in recent commit